### ISSUE TITLE
too many watchers

### ISSUE THREAD

**Issue by szexigexi** [720kb/angular-datepicker#25]
i added this plugin to my app and on a page with 6 date field the number of watchers increased by ~1800.  i used [this method](http://stackoverflow.com/a/18526757/3032512) to count the watchers.

**Comment by 45kb** [comment]
@szexigexi thank you for reporting, i need to test this deeply, if you have any idea feel free to PR, any help appreciated  :+1:

**Comment by wouldgo** [comment]
Already investigating in this... but I've very very little free time to spend :disappointed:

**Comment by afaqurk** [comment]
+1 I love this directive but had to remove it from our enterprise project because it registered 2000 watchers for 7 date fields on a very plain form.

Here is the output from [ng-stats](https://github.com/kentcdodds/ng-stats):
![](http://i.imgur.com/IGc5bTs.png)

First number is the # of watchers on the page and the 2nd number is the average time of digest. If I remove the 7 date fields, I get: 45 | 0.61

Somehow, this directive is creating hundreds of watchers.

**Comment by wouldgo** [comment]
Thanks to bubble this to my eyes again... I completely forgot this. Done some review and rearrange some behaviors now for the 3 datepickers page I have 386 watchers.

I hope to decrease the number much more, if you have some advice let us know.
Any help is appreciated.

Have a great day,
Dario.

**Comment by tjsail33** [comment]
could this be partially resolved by using an ng-if when the datepicker is not open so remove ALL of the datepicker watchers until they are needed?

Theoretically there will only be one datepicker open at a time, so this solution means you would have a max of 129 additional watchers on a page when utilizing this datepicker.

**Comment by dhaspden** [comment]
This is an issue for me as well.  I'm willing to lend a hand to get the watcher count down if you want but I don't know the project particularly well.

**Comment by leofachini** [comment]
Hi guys. I found something strange, that might help you.

In this [snippet](https://github.com/720kb/angular-datepicker/blob/master/src/js/angular-datepicker.js#L778-L785) bellow, look that on destroy this should remove a listener in this function 
[unregisterDataSetWatcher()](https://github.com/720kb/angular-datepicker/blob/master/src/js/angular-datepicker.js#L323-L338 )

```
       $scope.$on('$destroy', function unregisterListener() {

          unregisterDataSetWatcher(); //Pay attention to this method.
          thisInput.off('focus click focusout blur');
          angular.element(theCalendar).off('mouseenter mouseleave focusin');
          angular.element($window).off('click focus');
        });
      };
```
This is the unregosterDataSetWatcher snippet :
```
unregisterDataSetWatcher = $scope.$watch('dateSet', function dateSetWatcher(newValue) {

            if (newValue) {

              date = new Date(newValue);
              $scope.month = $filter('date')(date, 'MMMM');//december-November like
              $scope.monthNumber = Number($filter('date')(date, 'MM')); // 01-12 like
              $scope.day = Number($filter('date')(date, 'dd')); //01-31 like
              $scope.year = Number($filter('date')(date, 'yyyy'));//2014 like
              setDaysInMonth($scope.monthNumber, $scope.year);
              if ($scope.dateSetHidden !== 'true') {

                setInputValue();
              }
            }
```          

Everytime $destroy occurs this is registering another watch? Is that right?

**Comment by leofachini** [comment]
@wouldgo I might have a valid suggestion!

At the htmlTemplates, we could use :: (**One-time bindings** introduced in angular 1.3) to reduce the number of $watchers created by the bindable inside the ng-repeat.

Example:
```
generateDaysColumns = function generateDaysColumns() {

      return [
      '<div class="_720kb-datepicker-calendar-days-header">',
        '<div ng-repeat="d in daysInString">',  // <- Here we don't, because we need to track daysInString Array
          '{{::d}}', //< - Here we use the one-time biding
        '</div>',
      '</div>'
      ];
```

I have done some tests with ngStats and here is the outcome:

1 - **My environment**, not using angularjs-datepicker,  had a total amount of: 
![image](https://cloud.githubusercontent.com/assets/9139038/13863046/b5e7ee02-ec75-11e5-9514-ab9e9e5f0779.png) (37 watchers)

2 - **With** the angularjs-datepicker directive I had a total amount of:
![image](https://cloud.githubusercontent.com/assets/9139038/13863075/0adf66e2-ec76-11e5-975c-a997bdee7134.png) (166 watchers)


3 - **With proposed changes on the datepicker directive**, the total amount of:
![image](https://cloud.githubusercontent.com/assets/9139038/13863156/ace1d27c-ec76-11e5-99ef-eb31c278750a.png) (104 watchers)

Doing this we could reduce the total amount of watchers from 166 to 105. 61 less watchers or 36,74% less watchers . 

As far as I can tell, the calendar is working just fine with the changes.

**Comment by leofachini** [comment]
Another thing we could do is to remove as many as possible ng-class from our templates and manipulate it in our directive. This way we don't force extra data biding. By the way, the range of years that appears in the datepicker should/could be informed by the user(if the programmer didn't want to show all the 20 years range). 

I'm working on this to use in my project and if you guys approve I could do some pull request. 

C ya.

**Comment by leofachini** [comment]
Hello everyone, it's me again.

I've done my work in those watchers. I drop from **129 watchers** to **7 watchers** !!! :smile: 

Most of that can be acomplished doing the things that I've saied above.

Regards.

**Comment by dhaspden** [comment]
I would be very grateful for this to be merged in.  I use this library extensively and would really benefit from this!  Thanks for the hard work :smiley:!

**Comment by flick36** [comment]
@leofachini can you share the changes you made in your repo maybe? or make the PR please

**Comment by leofachini** [comment]
Hi guys! 
I can't do a PR rightnow. Nonetheless, [there are the changes that I've made in datepicker file.](https://github.com/leofachini/angular-datepicker/blob/master/src/js/angular-datepicker.js)

I use [Angular Style Guide by John Papa](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md) as reference to rewrite the directive. I strongly suggest you guys to read this guide, since it helped me a lot.

**Notice that I didn't do any changes in regards to the mobile watchers, neither tested. So use it on your own risk!**

Also I have made some css tricks and added some custom things.

There is a snippet about my less file with some css: 
```
/* ss-datepicker begin */
ss-datepicker a, [ss-datepicker] a, .ss-datepicker a{
  color:inherit;
  text-decoration:none;
}
ss-datepicker a:hover, [ss-datepicker] a:hover, .ss-datepicker a:hover{
  text-decoration:none;
}
ss-datepicker select, ss-datepicker select:focus, ss-datepicker select:hover,
.ss-datepicker select, .ss-datepicker select:focus, .ss-datepicker select:hover,
[ss-datepicker] select, [ss-datepicker] select:focus, [ss-datepicker] select:hover{
  width:100%;
   overflow: hidden;
   background:none;
   color:#fff;
   background-color: #138EFA;
   border-radius:2px;
   border: 0;
   margin-top:5px;
}
ss-datepicker, .ss-datepicker, [ss-datepicker],
._720kb-datepicker-calendar-header,
._720kb-datepicker-calendar-body,
._720kb-datepicker-calendar-days-header,
._720kb-datepicker-calendar-years-pagination-pages {
  font-family: Helvetica Neue;
  font-size: 13.5px;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  -ms-box-sizing: border-box;
  box-sizing: border-box;
  width: 100%;
  margin: 0 auto;
  float: left;
  clear: right;
  position: relative;
}

._720kb-datepicker-calendar-header-left {
    min-width: 15%;
    min-height: 1px;
}

._720kb-datepicker-calendar-day._720kb-datepicker-active {
  background: rgba(0, 0, 0, 0.36);
}

._720kb-datepicker-calendar-day:hover {
  background: rgba(0, 156, 255, 0.34);
}

._720kb-datepicker-calendar-header:nth-child(odd) {
    background: #DF5700;
}

._720kb-datepicker-calendar-days-header {
    background: rgba(223, 87, 0, 0.18);

    div {
        width: 14.28%;
    }
}

._720kb-datepicker-calendar-month span {
    color: #FFF;
}

._720kb-datepicker-default-button {
    color: #FFF;
}
/* ss-datepicker end */
```

And the HTML snippet:

```
<ss-datepicker date-format="dd/MM/yyyy"
			date-set="vm.abastecimento.data"
			how-many-years-to-show=8
			date-set-hidden
			date-refocus>
			<div class="form-group">
				<label for="data">Data Abastecimento</label>
				<div class="input-group">
					<input id="data"
						class="form-control"
						type="text"
						placeholder="Escolha uma data"
						ng-model="vm.abastecimento.data"
						required>
					<span class="input-group-addon" style="cursor: pointer">
						<i class="glyphicon glyphicon-arrow-down"></i>
					</span>
				</div>
				<div class="error-messages"
					ng-messages="vm.abastecimentoForm.data.$error"
					ng-if="(vm.abastecimentoForm.data.$touched && vm.abastecimentoForm.data.$dirty) || vm.showErrors"
					role="alert">
					<div ng-message="required">
						Você deve informar a data que ocorreu o abastecimento.
					</div>
				</div>
			</div>
		</ss-datepicker>
```

To sum up, I hope that this post might be of some help or use to you in order to identify where are the issues realted to the wathcers as code style to further enhacements or bug corrections as well.

See you guys! :sunglasses:

**Comment by evangrim** [comment]
> 

Still have these changes handy? Looks like the linked gist (containing the JS?) was removed @[leofachini](https://github.com/leofachini) from:
https://github.com/leofachini/angular-datepicker/blob/master/src/js/angular-datepicker.js

**Comment by leofachini** [comment]
Sorry @evangrim I guess that I’ve had deleted this  fork a while a go. Either way, the things that I’ve states here in the past are still valid. The outcome result was to fork this repo and use it as basis to create a custom date picker component inside your project.
